### PR TITLE
make clientid configurable

### PR DESCRIPTION
**Pull request by teichsta**
Hi,

this is a PR according to the discussion in #223. Please forgive that i never wrote C code so this one mainly is copy pasting code. I am also not sure about the variable assignment in L363.

It would be great if you could have a look and probably merge this one quickly as i am currently stuck and i hope to get your feedback on this soon.

Thanks a lot, Thomas E.-E.

### CODE REVIEW

File: src/ebusd/mqtthandler.cpp [hunk 1]
```diff
@@ -36,30 +36,32 @@ using std::dec;
 #define O_RETA (O_TOPI+1)
 #define O_JSON (O_RETA+1)
 #define O_IGIN (O_JSON+1)
+#define O_CLID (O_KEYF+1)
```
Comment (andig) [inline src/ebusd/mqtthandler.cpp]: O_KEYF+1 looks wrong and is a duplicate ;)

File: src/ebusd/mqtthandler.cpp [hunk 2]
```diff
@@ -151,6 +154,14 @@ static error_t mqtt_parse_opt(int key, char *arg, struct argp_state *state) {
     g_ignoreInvalidParams = true;
     break;
 
+  case O_CLID:  // --mqttclientid=clientid
+    if (arg == nullptr) {
```
Comment (andig) [inline src/ebusd/mqtthandler.cpp:158]: Just guessing, but maybe add 

    || arg[0] == 0

to check for 0 length, same as O_TOPI?

**Comment by teichsta** [comment]
@john30 any chance to look into this PR?

**Comment by john30** [comment]
Not this week

**Comment by teichsta** [comment]
> Not this week

ok, got the code running on my machine, so no worries. I also added an option to set MQTT protocol version to 3.1.1. With this setup i am able to connect ebusd to Google Cloud IoT Core. It would be great to add a function to generate the JWT token to improve the authentication bit.

**Comment by andig** [comment]
> It would be great to add a function to generate the JWT token to improve the authentication bit.

Why should that be ebusd responsibility?

**Comment by teichsta** [comment]
> Why should that be ebusd responsibility?

because ebusd (in this case) acts as a MQTT client which needs to authenticate against a broker. There are several authentication mechanisms and one of which is JWT. Another is user/password which also is supported.

**Comment by john30** [comment]
does not work this way for backwards compatibility

**Comment by andig** [comment]
> does not work this way for backwards compatibility

@john30: was this comment potentially meant for https://github.com/john30/ebusd/issues/219?

**Comment by john30** [comment]
@andig no for this one, as Travis CI build has failed

**Comment by andig** [comment]
@teichsta could you take a look? MQTT 3.1.1 doesn't compile:

    mqtthandler.cpp: In constructor ‘ebusd::MqttHandler::MqttHandler(ebusd::UserInfo*, ebusd::BusHandler*, ebusd::MessageMap*)’:
    mqtthandler.cpp:398:33: error: ‘MQTT_PROTOCOL_V311’ was not declared in this scope
        int mqttProtocolVersion = MQTT_PROTOCOL_V311;
                                    ^
    mqtthandler.cpp:400:43: error: ‘MOSQ_OPT_PROTOCOL_VERSION’ was not declared in this scope
        if (mosquitto_opts_set(m_mosquitto, MOSQ_OPT_PROTOCOL_VERSION, &mqttProtocolVersion) != MOSQ_ERR_SUCCESS) {
                                            ^
    mqtthandler.cpp:400:90: error: ‘mosquitto_opts_set’ was not declared in this scope
        if (mosquitto_opts_set(m_mosquitto, MOSQ_OPT_PROTOCOL_VERSION, &mqttProtocolVersion) != MOSQ_ERR_SUCCESS) {
                                
Maybe split that into a separate PR?

**Comment by teichsta** [comment]
> Maybe split that into a separate PR?

yes, reverted this MQTT v3.1.1 switch and created another PR (see #227) for this!

File: src/ebusd/mqtthandler.cpp [hunk 3]
```diff
@@ -36,30 +36,32 @@ using std::dec;
...
+  {"mqtttopic",    O_TOPI, "TOPIC",       0,
    "Use MQTT TOPIC (prefix before /%circuit/%name or complete format) [ebusd]", 0 },
-  {"mqttretain",  O_RETA, nullptr,       0, "Retain all topics instead of only selected global ones", 0 },
-  {"mqttjson",    O_JSON, nullptr,       0, "Publish in JSON format instead of strings", 0 },
-  {"mqttignoreinvalid", O_IGIN, nullptr, 0,
+  {"mqttretain",   O_RETA, nullptr,       0, "Retain all topics instead of only selected global ones", 0 },
+  {"mqttjson",     O_JSON, nullptr,       0, "Publish in JSON format instead of strings", 0 },
+  {"mqttignoreinvalid", O_IGIN, nullptr,  0,
    "Ignore invalid parameters during init (e.g. for DNS not resolvable yet)", 0 },
+  {"mqttclientid", O_CLID, "CLIENTID",    0, "Set the MQTT CLIENTID, defaults to PACKAGE_NAME_PACKAGE_VERSION_PID", 0 },
```
Comment (john30) [inline src/ebusd/mqtthandler.cpp:58]: rather:
"Set the MQTT CLIENTID [" PACKAGE_NAME "_" PACKAGE_VERSION "_<pid>]"

File: src/ebusd/mqtthandler.cpp [hunk 4]
```diff
@@ -69,6 +71,7 @@ static const char* g_host = "localhost";  //!< host name of MQTT broker [localho
 static uint16_t g_port = 0;               //!< optional port of MQTT broker, 0 to disable [0]
 static const char* g_username = nullptr;  //!< optional user name for MQTT broker (no default)
 static const char* g_password = nullptr;  //!< optional password for MQTT broker (no default)
+static const char* g_clientId = nullptr;  //!< optional clientid, defaults to PACKAGE_NAME_PACKAGE_VERSION_PID
```
Comment (john30) [inline src/ebusd/mqtthandler.cpp:74]: rather:
optional clientid [PACKAGE_NAME '_' PACKAGE_VERSION '_' <pid>]

File: src/ebusd/mqtthandler.cpp (hunk 2 above)
Comment (john30) [inline src/ebusd/mqtthandler.cpp:158]: also check length > 0

**Review by john30 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**